zfs: Relate import services to zfs-import.target instead of local-fs

ElvishJerricco opened the pull request:
Being wanted by and ordered before local-fs.target isn't strictly correct. And in systemd initrd, it's very incorrect because local-fs.target is for the initrd file system, not the real root file system.

###### Description of changes

This changes the import services to not be ordered before or wanted by local-fs.target.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

ElvishJerricco committed:
zfs: Relate import services to zfs-import.target instead of local-fs

Being wanted by and ordered before local-fs.target isn't strictly
correct. And in systemd initrd, it's very incorrect because
local-fs.target is for the initrd file system, not the real root file
system.

flokli commented on nixos/modules/tasks/filesystems/zfs.nix:
```diff
@@ -582,6 +582,8 @@ in
             prefix = "/sysroot";
           }) rootPools);
+          targets.zfs-import.wantedBy = [ "zfs.target" ];
+          targets.zfs.wantedBy = [ "initrd.target" ];
```
Is zfs.target reached in case pools are slightly degraded, but still mountable?

ElvishJerricco replied:
Hm, I suppose I changed `wantedBy = ... ["local-fs.target"]` to `requiredBy = ... ["zfs-import.target"]`. So if any pool fails to import, then `zfs-import.target` will fail. But that is only *wanted* by `zfs.target`, so that would still be reached. Also, our `zfs-import-${pool}.service` *does* attempt to import pools degraded if possible.

RaitoBezarius replied:
I'm not sure about having a generic name like `zfs.target`, could have something more descriptive?

ElvishJerricco replied:
@RaitoBezarius That's just the name of the upstream unit that we get from `systemd.packages = [ cfgZfs.package ];`. I don't think we can change it, for compatibility's sake.

RaitoBezarius replied:
Got it

ElvishJerricco replied:
@flokli Is there a problem with this?

flokli replied:
> @flokli Is there a problem with this?

I don't know, I mostly wanted to make sure we can still boot even with a degraded pool. You don't want your machine to be unbootable just because one disk doesn't come back after a reboot.

If you're certain this still works, feel free to mark this thread as resolved :-)

ElvishJerricco replied:
Well the logic for handling a degraded pool is in the import service itself, so I don't think that concern has to do with the target ordering.

ElvishJerricco commented:
I've rebased on nixos-unstable-small, and I've also added a commit that orders `zfs-import-${pool}.service` after `cryptsetup.target`, because otherwise the service can time out while waiting on the user to unlock a LUKS device. If desired, I can split that out into a separate PR. But it seemed close enough related to this one to me.

someplaceguy commented:
Initially I wasn't sure about removing the `local-fs.target` dependency/ordering, but after looking into it more deeply, it makes sense to me and I don't think it should cause any major issues.

I've tested this on 3 systems, each with one ZFS root pool and one ZFS boot pool, two of which having both pools encrypted with LUKS, and all of them seem to have booted fine at least once.

None of these 3 systems are running systemd in initrd, though. I tried enabling that option on one of my systems but it became unusable due to #244737.

Perhaps it should be mentioned that this PR fixes issues when using both ZFS and systemd in initrd, which helps solving issue #74281, so I hope this gets merged soon.

Thanks!

ElvishJerricco commented:
The only potential issue I see with this PR is that anyone who has `/etc/crypttab` entries dependent on zvols will have their setup break with this new ordering. Or similar with `boot.initrd.luks.devices` entries and a zvol from a pool imported in systemd stage 1.

someplaceguy commented:
> The only potential issue I see with this PR is that anyone who has `/etc/crypttab` entries dependent on zvols will have their setup break with this new ordering. Or similar with `boot.initrd.luks.devices` entries and a zvol from a pool imported in systemd stage 1.

Ah, I hadn't considered that possibility.

That seems like a hard issue to solve unless all the dependencies between LUKS devices and ZFS pools are either manually specified or somehow automatically detected, and respected.

I'm saying this since, of course, you can have a LUKS device with a ZFS pool sitting on top of it, which contains a ZVOL which in turn contains another LUKS device. So any kind of static ordering would not work in this situation, it does seem like it would require fine-grained dependencies...

I mean, I think this could work instead by either importing a ZFS pool or opening LUKS devices in a fault-tolerant way and then just run these services a couple of times until everything succeeds, but that seems really hacky and error-prone...

someplaceguy commented:
> I'm saying this since, of course, you can have a LUKS device which contains a ZFS pool, which in turn contains a ZVOL which in turn contains another LUKS device. So any kind of static ordering would not work in this situation, it does seem like it would require fine-grained dependencies instead...

By the way, this is not just a theoretical issue, I just remembered that I do have exactly such a configuration on one of my systems for legitimate reasons...

ElvishJerricco commented:
The real solution is reviving https://github.com/openzfs/zfs/pull/4943 and making ZFS pools import based on device rediness via udev rules. [btrfs does this](https://github.com/systemd/systemd/blob/main/rules.d/64-btrfs.rules.in) pretty nicely. Basically, the device unit that a btrfs mount unit depends on doesn't become active until the btrfs pool has all its disks and is "ready".

The closed ZFS PR created a `/dev/zpool/poolname/ready` device representing the readiness of pools. A zfs import service could then just depend on that the same way mount units depend on regular device units. We could even add a `ready-degraded` device, so that you could have:

```
Wants=dev-zpool-tank-ready.device
Requires=dev-zpool-tank-ready-degraded.device
After=dev-zpool-tank-ready.device dev-zpool-tank-ready-degraded.device
```

So that it times out waiting for real readiness but outright won't start if it times out waiting for degraded readiness.

Regardless, all of this is out of scope for this PR.

ElvishJerricco commented:
I'm going to remove the `cryptsetup` change. We can solve that elsewhere.

someplaceguy commented:
> I'm going to remove the `cryptsetup` change. We can solve that elsewhere.

Can you elaborate on how that is or will be solved? I'd like to replace your commit with the proper fix on my systems.

Thanks!
